src/redisearch_rs/slots_tracker/src/slots_tracker.rs

```diff
@@ -0,0 +1,932 @@
+    /// The internal version is always Stable.
+    pub const fn get_version(&self) -> Version {
+        // Internal invariant: self.version is always Stable
+        self.version
+    }
+
+    /// Increments the version counter using wrapping arithmetic.
+    ///
+    /// Internal invariant: This is always called on Stable versions.
+    fn increment_version(&mut self) {
+        // version is always Stable internally
+        self.version = self.version.increment();
+    }
+
+    /// Sets the local responsibility slot ranges.
+    ///
+    /// This function replaces the `local` set with the provided ranges.
+    /// It also removes the given slots from `fully_available` and `partially_available`.
+    /// If the new ranges are identical to the current local slots, no change occurs and version
+    /// is not incremented.
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges. Must be sorted and normalized (no overlaps, no adjacent ranges).
+    pub fn set_local_slots(&mut self, ranges: &[SlotRange]) {
+        // Check if the ranges are already equal (no change needed)
+        if self.local == ranges {
+            return;
+        }
+
+        // Update local slots and remove from other sets
+        self.local = SlotSet::from_ranges(ranges);
+        self.fully_available.remove_ranges(ranges);
+        self.partially_available.remove_ranges(ranges);
+        self.increment_version();
+    }
+
+    /// Marks the given slot ranges as partially available.
+    ///
+    /// This function updates the `partially_available` set by adding the provided ranges.
+    /// It also removes the given slots from `local` and `fully_available`, and
+    /// increments the `version` counter (fully available or not available slots are now partially available).
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges. Must be sorted and normalized (no overlaps, no adjacent ranges).
+    pub fn mark_partially_available_slots(&mut self, ranges: &[SlotRange]) {
+        // Update partially available slots and remove from other sets
+        self.partially_available.add_ranges(ranges);
+        self.local.remove_ranges(ranges);
+        self.fully_available.remove_ranges(ranges);
+        self.increment_version();
+    }
+
+    /// Promotes slot ranges to local ownership.
+    ///
+    /// This function adds the provided ranges to `local` and removes them from `partially_available`.
+    /// Does NOT modify `fully_available` and does NOT increment the version counter
+    /// (the version was already bumped when slots became partially available, and while partially
+    /// available slots exist, `check_availability` returns unstable/unavailable anyway).
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges. Must be sorted and normalized (no overlaps, no adjacent ranges).
+    pub fn promote_to_local_slots(&mut self, ranges: &[SlotRange]) {
+        debug_assert!(matches!(
```

Comment:
can we assert then that fully_available does not contain them? This would give more certainty that the dependency call (version counter increase) is satisfied.